### DISCUSSION TITLE
Support new events using Java Agent Toolkit

### DISCUSSION THREAD

**Discussion by nisiyong** [apache/skywalking#10079] [Ideas]
Currently, SkyWalking can auto-detect the topology by analyzing the segment data. It means if we want to detect some components, we must trace them. In our production practices,  we use `skywalking-java` to instrument our Java applications, some plugins will be excluded, like Redis. 

There will be a lot of spans when we trace the Redis operations, because each Redis operation will generate a span, and most business applications will use Redis operation in the for-loops, it is hard to change the business logic, once we trace the Redis operations, the span max size (default 300) is very easy exceeded, then those other important spans will be ignored. When diagnosis production I prefer using Redis metrics, rather than Redis tracing.

Now I want to auto-detect the topology in a lightweight, I think the [Event](https://skywalking.apache.org/docs/main/next/en/concepts-and-designs/event/ ) could meet it. We could define the new events, like `Connect`, `Disconnect`, `Register`, `Unregister`, `Subscribe`, `Unsubscribe`, etc. Then we could associate the server metrics to this service or instances.

Currently, the Java Agent Toolkit does not support the event report, I wanna implement it and support the new events. Do you recommend doing this?  And any suggestions? @wu-sheng @kezhenxu94

**Comment by wu-sheng** [comment]
Event doesn't mean an event of processes, it is an concept of events for service itself, such as booting, upgrade, config update.

**Reply by nisiyong** [reply]
Won't it support the instance level in future? Instances have events, too. If we use java-agent to do it, it may do a lot of interesting things.

**Reply by wu-sheng** [reply]
It supports the instance, but you are abusing the API.

https://github.com/apache/skywalking-data-collect-protocol/blob/b561375576d97a061e4bc3bf30f0f1ae66e2a436/event/Event.proto#L82-L86

**Reply by nisiyong** [reply]
Thanks, it indeed supports instance and endpoint event. But I am confuse why you say I am abusing the API, in my understanding, instance also have events, like Start, Stop, Restart, Connect, Disconnect, Register, Unregister, Subscribe, Unsubscribe, etc. Could you give more explain?

**Reply by wu-sheng** [reply]
You asked for topology from events, that is an abuse. Event is not for aggregation.

**Reply by nisiyong** [reply]
The events just record what service or instance happens. If the community thinks detecting topology from events is abuse or redundant, maybe we could analyze events personally.

But another thing is the java-agent lacks Event API currently, do we need to implement it?

**Reply by wu-sheng** [reply]
Detecting topology should be built by meter APIs if you need. But not an event. An event is a one-time execution, there is no benefit to analysis topology from it. 

> If the community thinks detecting topology from events is abuse or redundant, maybe we could analyze events personally.

You are free to do anything in the private fork, that is totally fine. I keep discussing with you this idea because you sent this proposal to the community.

**Reply by nisiyong** [reply]
Okay, thanks for your suggestion. And do we have any plan with Java-Agent Event API?

**Reply by wu-sheng** [reply]
If you could provide a use case about what events an agent or a service should send, I think it is fine to add one.

**Reply by nisiyong** [reply]
Okay, thanks.

**Comment by wu-sheng** [comment]
MAL has provided topology generate logic, once your meter includes both sides as tags.
